### ISSUE TITLE
Error in linking CXX executable during installation

### ISSUE THREAD

**Issue by asenabouth** [statgen/popscle#7]
Hello,

I'm having some issues with installing popscle. First thing, I'm trying to set it up on a shared compute cluster where I don't have admin privileges. I've been trying to manage my software via Anaconda.

Here is my conda environment:

```bash
(popscle) -bash-4.1$ conda list --name popscle
# packages in environment at /share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle:
#
# Name                    Version                   Build  Channel
bzip2                     1.0.6             h14c3975_1002    conda-forge
ca-certificates           2019.3.9             hecc5488_0    conda-forge
curl                      7.64.1               hf8cf82a_0    conda-forge
htslib                    1.8                  h57f8fdf_2    bioconda
krb5                      1.16.3            h05b26f9_1001    conda-forge
libcurl                   7.64.1               hda55be3_0    conda-forge
libdeflate                1.2                  h14c3975_0    bioconda
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libgcc-ng                 8.2.0                hdf63c60_1    anaconda
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              8.2.0                hdf63c60_1    anaconda
libzip                    1.5.1             h2f06484_1003    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
openssl                   1.1.1b               h14c3975_1    conda-forge
tk                        8.6.9             h84994c4_1001    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
zlib                      1.2.11            h14c3975_1004    conda-forge
```
I initially encountered similar issues to issue #5 so I followed the steps to resolve it by installing htlib 1.8 and replacing `stdhist.h` with `cstdint`. I had to modify the CXX make flag as follows 
to get it working though:

```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -pthread -std=gnu++11")
```
This was the resulting build message:

```bash
-- The C compiler identification is GNU 4.8.2
-- The CXX compiler identification is GNU 4.8.2
-- Check for working C compiler: /opt/rh/devtoolset-2/root/usr/bin/cc
-- Check for working C compiler: /opt/rh/devtoolset-2/root/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /opt/rh/devtoolset-2/root/usr/bin/c++
-- Check for working CXX compiler: /opt/rh/devtoolset-2/root/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- /share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/include
-- /share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a
-- Found BZip2: /usr/lib64/libbz2.so (found version "1.0.5") 
-- Looking for BZ2_bzCompressInit in /usr/lib64/libbz2.so
-- Looking for BZ2_bzCompressInit in /usr/lib64/libbz2.so - found
-- Configuring done
-- Generating done
-- Build files have been written to: /share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/include/popscle/build
```

I was able to compile the code in the package, but it fails at the linking step. Here are the messages:

```
Linking CXX executable ../bin/popscle
/opt/rh/devtoolset-2/root/usr/libexec/gcc/x86_64-redhat-linux/4.8.2/ld: /share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(plugin.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
/opt/rh/devtoolset-2/root/usr/libexec/gcc/x86_64-redhat-linux/4.8.2/ld: note: 'dlclose@@GLIBC_2.2.5' is defined in DSO /lib64/libdl.so.2 so try adding it to the linker command line
/lib64/libdl.so.2: could not read symbols: Invalid operation
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/popscle] Error 1
make[1]: *** [CMakeFiles/popscle.dir/all] Error 2
```

I had set LIBRARY_PATH and LD_LIBRARY_PATH in my .bashrc hoping that would help. It did not. I would appreciate any help on how to proceed.

Thanks!

PS: I was able to set up and run demuxlet on this environment, but I would like to try the new version.

**Comment by Griffan** [comment]
Could you try to add these flags to CMAKE_CXX_FLAGS:

set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -pthread -std=gnu++11 -ldl")

**Comment by asenabouth** [comment]
That helped. It was able to progress but ended with this error:

Linking CXX executable ../bin/popscle
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `bgzf_compress':
bgzf.c:(.text+0x1b55): undefined reference to `libdeflate_alloc_compressor'
bgzf.c:(.text+0x1b94): undefined reference to `libdeflate_deflate_compress'
bgzf.c:(.text+0x1bc8): undefined reference to `libdeflate_free_compressor'
bgzf.c:(.text+0x1bed): undefined reference to `libdeflate_free_compressor'
bgzf.c:(.text+0x1c39): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `bgzf_uncompress':
bgzf.c:(.text+0x1eb4): undefined reference to `libdeflate_alloc_decompressor'
bgzf.c:(.text+0x1f0c): undefined reference to `libdeflate_deflate_decompress'
bgzf.c:(.text+0x1f1b): undefined reference to `libdeflate_free_decompressor'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `inflate_block':
bgzf.c:(.text+0x1fd3): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(cram_io.o): In function `itf8_decode_crc':
cram_io.c:(.text+0x18a7): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x18ff): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x197d): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x1a21): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(cram_io.o):cram_io.c:(.text+0x1af4): more undefined references to `libdeflate_crc32' follow
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/popscle] Error 1
make[1]: *** [CMakeFiles/popscle.dir/all] Error 2
make: *** [all] Error 2

**Comment by hyunminkang** [comment]
You need to include the directiory containing libhts.a in LD_LIBRARY_PATH
-----------------------------------------------------
Hyun Min Kang, Ph.D.
Associate Professor of Biostatistics
University of Michigan, Ann Arbor
Email : hmkang@umich.edu

**Comment by asenabouth** [comment]
Directory with libhts.a has been added to library path.

I updated libdeflate via anaconda. This is the current environment:

```
# Name                    Version                   Build  Channel
bzip2                     1.0.6             h14c3975_1002    conda-forge
ca-certificates           2019.3.9             hecc5488_0    conda-forge
curl                      7.64.1               hf8cf82a_0    conda-forge
htslib                    1.8                  h57f8fdf_2    bioconda
krb5                      1.16.3            h05b26f9_1001    conda-forge
libcurl                   7.64.1               hda55be3_0    conda-forge
libdeflate                1.2                  h14c3975_0    bioconda
libedit                   3.1.20170329      hf8c457e_1001    conda-forge
libgcc-ng                 8.2.0                hdf63c60_1    anaconda
libssh2                   1.8.2                h22169c7_2    conda-forge
libstdcxx-ng              8.2.0                hdf63c60_1    anaconda
libzip                    1.5.1             h2f06484_1003    conda-forge
ncurses                   6.1               hf484d3e_1002    conda-forge
openssl                   1.1.1b               h14c3975_1    conda-forge
tk                        8.6.9             h84994c4_1001    conda-forge
xz                        5.2.4             h14c3975_1001    conda-forge
zlib                      1.2.11            h14c3975_1004    conda-forge
```

Make results in these errors:
```
Linking CXX executable ../bin/popscle
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `bgzf_compress':
bgzf.c:(.text+0x1b55): undefined reference to `libdeflate_alloc_compressor'
bgzf.c:(.text+0x1b94): undefined reference to `libdeflate_deflate_compress'
bgzf.c:(.text+0x1bc8): undefined reference to `libdeflate_free_compressor'
bgzf.c:(.text+0x1bed): undefined reference to `libdeflate_free_compressor'
bgzf.c:(.text+0x1c39): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `bgzf_uncompress':
bgzf.c:(.text+0x1eb4): undefined reference to `libdeflate_alloc_decompressor'
bgzf.c:(.text+0x1f0c): undefined reference to `libdeflate_deflate_decompress'
bgzf.c:(.text+0x1f1b): undefined reference to `libdeflate_free_decompressor'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(bgzf.o): In function `inflate_block':
bgzf.c:(.text+0x1fd3): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(cram_io.o): In function `itf8_decode_crc':
cram_io.c:(.text+0x18a7): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x18ff): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x197d): undefined reference to `libdeflate_crc32'
cram_io.c:(.text+0x1a21): undefined reference to `libdeflate_crc32'
/share/ClusterShare/software/contrib/annsen/anaconda3/envs/popscle/lib/libhts.a(cram_io.o):cram_io.c:(.text+0x1af4): more undefined references to `libdeflate_crc32' follow
collect2: error: ld returned 1 exit status
make[2]: *** [../bin/popscle] Error 1
make[1]: *** [CMakeFiles/popscle.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by Griffan** [comment]
Try to add -lz in CXX flags

**Comment by jonathonl** [comment]
These link errors are for libdeflate not libz. It looks like htslib was built with the `--with-libdeflate` configure option, so every executable that links to this build of htslib will need to also link against libdeflate (`CMAKE_CXX_FLAGS="-ldeflate"`).

**Comment by jonathonl** [comment]
On second thought, it's possible that using CMAKE_CXX_FLAGS may not work since link order matters with static libs. In which case you'll need to add deflate to target_link_libraries() in the CMakeLists.txt file.

**Comment by Griffan** [comment]
I agree. And this related issue(https://github.com/samtools/htslib/issues/688) might help.

**Comment by asenabouth** [comment]
We finally have success! Thank you all for the help - the comments regarding libdeflate were particularly helpful as I was using the bioconda version of htslib. This version was compiled with libdeflate.

I had to install libdeflate from source. As I don't have sudo permissions on my HPC, I had to add its installation paths to both LIBRARY_PATH and LD_LIBRARY_PATH in my .bashrc file.

My final CMAKE_CXX_FLAG looks like this:

```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3 -pthread -std=gnu++11 -ldl -ldeflate")
```

I didn't need to edit the target_link_libraries() variable in the end.
